daemon/containerd/image_push.go

```diff
@@ -2,14 +2,247 @@ package containerd
+			}), h)
+	}
+	return remotes.PushContent(ctx, pusher, target, store, limiter, platforms.All, addChildrenToJobs)
+}
+
+// prepareMissing will walk the target descriptor recursively and handles
+// missing content which to make the target descriptor possible to be pushed.
+//
+// If target references any missing content it will be fetched from the source
+// repository unless it can be cross-repo mounted.
+//
+// Manifests/configs can't be mounted and need to be downloaded - it shouldn't
+// be an issue though, because they are relatively small json files.
+// If we are pushing to the same registry then layer blobs can be
+// cross-repo mounted and won't be downloaded.
+func (i *ImageService) prepareMissing(ctx context.Context, store content.Store, jobs *jobs, resolver remotes.Resolver,
+	target ocispec.Descriptor, targetRef reference.Named,
+	progress *pushProgress, limiter *semaphore.Weighted,
+) (map[digest.Digest]distributionSource, error) {
+	mountableBlobs := map[digest.Digest]distributionSource{}
+	source, err := getDigestSource(ctx, store, target.Digest)
+
+	if err != nil {
+		if !errdefs.IsNotFound(err) {
+			return nil, err
+		}
+		logrus.WithField("target", target).Debug("distribution source label not found")
+		return mountableBlobs, nil
+	}
+
+	handler := func(ctx context.Context, desc ocispec.Descriptor) ([]ocispec.Descriptor, error) {
+		_, err := store.Info(ctx, desc.Digest)
+		if err != nil {
+			if !cerrdefs.IsNotFound(err) {
+				return nil, errdefs.System(errors.Wrapf(err, "failed to get metadata of content %s", desc.Digest.String()))
+			}
+
+			ref, err := source.GetReference(desc.Digest)
+			if err != nil {
+				return nil, errdefs.InvalidParameter(errors.Wrapf(err, "failed to get reference to content %s", desc.Digest.String()))
+			}
+
+			jobs.Add(desc)
+			if canBeMounted(desc, targetRef, i.registryService.IsInsecureRegistry, source) {
+				dgst := desc.Digest
+				mountableBlobs[dgst] = source
+				progress.addMountable(dgst)
+				return nil, nil
+			} else {
+				fetcher, err := resolver.Fetcher(ctx, ref.String())
+				if err != nil {
+					return nil, errdefs.InvalidParameter(errors.Wrapf(err, "failed to create fetcher for %s", ref.String()))
+				}
+
+				fetchHandler := remotes.FetchHandler(store, fetcher)
+				_, err = fetchHandler(ctx, desc)
+				if err != nil {
+					return nil, errdefs.System(errors.Wrapf(err, "failed to fetch descriptor %s (type %s)", desc.Digest.String(), desc.MediaType))
+				}
+			}
+		}
+
+		return containerdimages.Children(ctx, store, desc)
+	}
+
+	err = containerdimages.Dispatch(ctx, containerdimages.HandlerFunc(handler), limiter, target)
+	if err != nil {
+		return nil, err
+	}
+
+	// Set GC labels recursively for all downloaded blobs.
+	setLabels := containerdimages.SetChildrenLabels(store, containerdimages.ChildrenHandler(store))
+	err = containerdimages.Dispatch(ctx, setLabels, nil, target)
+	if err != nil {
+		logrus.WithError(err).Warn("set children gc labels failed")
+	}
```

Comment:
Not sure about this one - it persist downloaded content if it couldn't be cross repo mounted.
In case of pushing shallowly pulled image to a different registry, this will effectively unshallow an image and make it possible to use other downloaded platform.
On the other hand, deleting them just after we finish push is wasteful.

This should probably be an opt-in (once we have a CLI flag for that).